libsql/internal/http/hranaV2/hranaV2.go

```diff
@@ -286,7 +287,119 @@ func sendPipelineRequest(ctx context.Context, msg *hrana.PipelineRequest, url st
 	return result, false, nil
 }
 
+func (h *hranaV2Conn) executeMsg(ctx context.Context, query string, msg *hrana.PipelineRequest) (*hrana.PipelineResponse, error) {
+	result, err := h.sendPipelineRequest(ctx, msg, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute SQL: %s\n%w", query, err)
+	}
+
+	for _, r := range result.Results {
+		if r.Error != nil {
+			return nil, fmt.Errorf("failed to execute SQL: %s\n%s", query, r.Error.Message)
+		}
+		if r.Response == nil {
+			return nil, fmt.Errorf("failed to execute SQL: %s\n%s", query, "no response received")
+		}
+	}
+	return result, nil
+}
+
+func (h *hranaV2Conn) executeInChunks(ctx context.Context, query string, wantRows bool) (*hrana.PipelineResponse, error) {
+	iterator := sqliteparserutils.CreateStatementIterator(query)
+
+	stmt, _, isEOF := iterator.Next()
+	// Skip all empty statements
+	for {
+		if stmt != "" {
+			break
+		}
+		stmt, _, isEOF = iterator.Next()
+		if isEOF {
+			break
+		}
+	}
+	if isEOF {
+		msg := &hrana.PipelineRequest{}
+		executeStream, err := hrana.ExecuteStream(stmt, nil, wantRows)
+		if err != nil {
+			return nil, fmt.Errorf("failed to execute SQL: %s\n%w", query, err)
+		}
+		msg.Add(*executeStream)
+		return h.executeMsg(ctx, query, msg)
+	}
+
+	sqls := make([]string, 4096)
+
+	batch := &hrana.Batch{Steps: make([]hrana.BatchStep, len(sqls))}
+	msg := &hrana.PipelineRequest{}
+	msg.Add(hrana.StreamRequest{Type: "batch", Batch: batch})
+	for idx := range batch.Steps {
+		batch.Steps[idx].Stmt.WantRows = wantRows
+	}
+
+	result := &hrana.PipelineResponse{}
+
+	sqls[0] = "BEGIN"
+	batch.Steps[0].Stmt.Sql = &sqls[0]
+	nextIdx := 1
+	first := true
+	for {
+		if stmt != "" {
+			sqls[nextIdx] = stmt
+			batch.Steps[nextIdx].Stmt.Sql = &sqls[nextIdx]
+			nextIdx += 1
+			if nextIdx == len(sqls) {
+				res, err := h.executeMsg(ctx, query, msg)
+				if err != nil {
+					h.closeStream()
+					return nil, err
+				}
+				result.Baton = res.Baton
+				result.BaseUrl = res.BaseUrl
+				idx := 0
+				if first {
+					idx = 1
+				}
+				for ; idx < len(res.Results); idx++ {
+					result.Results = append(result.Results, res.Results[idx])
+				}
+				first = false
+				nextIdx = 0
+			}
+		}
+		if isEOF {
+			break
+		}
+		stmt, _, isEOF = iterator.Next()
+	}
+	sqls[nextIdx] = "COMMIT"
+	batch.Steps[nextIdx].Stmt.Sql = &sqls[nextIdx]
+	nextIdx += 1
+	if nextIdx < len(sqls) {
+		batch.Steps = batch.Steps[:nextIdx+1]
+	}
+	res, err := h.executeMsg(ctx, query, msg)
+	if err != nil {
+		h.closeStream()
+		return nil, err
+	}
+	result.Baton = res.Baton
+	result.BaseUrl = res.BaseUrl
+	idx := 0
+	if first {
+		idx = 1
+	}
+	for ; idx+1 < len(res.Results); idx++ {
+		result.Results = append(result.Results, res.Results[idx])
+	}
+
+	return result, nil
+}
+
 func (h *hranaV2Conn) executeStmt(ctx context.Context, query string, args []driver.NamedValue, wantRows bool) (*hrana.PipelineResponse, error) {
+	if len(args) == 0 && len(query) > 20*1024*1024 && !h.schemaDb {
```

Comment:
Why we we need this limit at all? Why not just always use chunking but with an optimization that if everything fits into single chunk - we just send single batch request without transaction?

Reply:
I don't want to affect all existing users. This is an optimization for a very special case and I'd rather keep the impact of the change as small as possible as it is not clear to me if chunking is always a desired behaviour for every workload.

Reply:
Maybe then we better to add new explicitly method which will have suitable signature (no args) and explicit name?

Reply:
Ok, new method will require more work and also questionable decision. 

But I don't like that we are trying to "mask" the potential issue. If we are afraid that chunking can break something for some users - we better to be more explicit about this. If there is noting to be afraid - we should make this logic unconditional in order to simplify code and catch bugs earlier.

Reply:
The problem is that we're using the driver through a standarized database/sql so adding custom methods is not easy

Reply:
There's plenty to be worried about as the new approach uses interactive transaction that blocks all the writes. It makes the whole thing tricky. It's probably worth it for really big batches as sqld wouldn't be able to handle them anyway but for smaller but still big batches like 10M it is safer to just send them in a single shot.

Reply:
> interactive transaction that blocks all the writes

Ok, I see. But what if the client will crash in the middle? Will writes be blocked forever? Or they will unblock after some timeout?

Reply:
They will unblock after a timeout.

Reply:
Ok, now I understand concern. Let's leave this threshold for applying chunking (but maybe move to the constant?)

Thanks for the explanation!

Reply:
No worries. I will turn it into a constant.